packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2632,7 +2632,12 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
 
     public async uploadBlob(blob: ArrayBufferLike): Promise<IFluidHandle<ArrayBufferLike>> {
         this.verifyNotClosed();
-        return this.blobManager.createBlob(blob);
+        const previousDirtyState = this.isDirty;
+        this.emitDirtyDocumentEvent = true;
```

Comment:
I do not understand what this code does or tries to do. I want to point out that there could be many calls to this API in parallel, so any modification of global state that does not accumulate (i.e. counters vs. boolean) is very likely a bug. Plus changes in state that are not revered at the end of the process (this.emitDirtyDocumentEvent is changed to true and never reset - not sure what's the thinking here)

Reply:
For some reason I believed the emitDirtyDocumentEvent was changed during updateDocumentDirtyState but I see it does not. That could be easily fixed by preserving the previous state. I agree asynchoronous calls are a problem in this implementation. What I could do is to check BlobManager.pendingBlobs size in line 1111 (constructor) but not sure we will be able to immediately flag the container as dirty like I'm doing it in this implementation.